catlearn/relation_cache.py

```diff
@@ -281,6 +315,109 @@ def flush(self) -> None:
+        for arrow in self.arrows():
+            src = arrow[0]
+            tar = arrow[-1]
+            other_scores = (
+                scores[arr] for arr in self.arrows(src, tar)
+                if arr is not arrow)
+
+            arrow_utility = max(
+                scores[arrow] - max(other_scores, default=-inf), 0.)
+            for idx in range(len(arrow)):
+                utility[arrow[idx:(idx + 1)]] += arrow_utility
+
+        # identify worst relation
+        candidate_arrows = (arr for arr, val in utility.items() if val < inf)
+        to_remove = min(
+            candidate_arrows, key=lambda arr: utility[arr], default=None)
+        return to_remove
+
+    def prune_relations(
+        self, nb_to_keep: int) -> FrozenSet[
+            CompositeArrow[NodeType, ArrowType]]:
+        """
+            Remove relations with a low score in the cache, and keep only
+            nb_to_keep relations of each order.
+            Only remove 1st order arrows. Arrows which are removed are those
+            with the lowest score relative to other arrows.
+
+            Returns the list of pruned relations
+        """
+        pruned = set()
+        while True:
+            relation = (
+                None if len(self) <= nb_to_keep
+                else self._get_worst_relation())
+            if relation is None:
+                return frozenset(pruned)
+            del self[relation]
+            pruned.add(relation)
+
+    def build_composites(
+        self, max_arrow_number,
+    ) -> Tuple[
+        FrozenSet[CompositeArrow[NodeType, ArrowType]],
+        FrozenSet[CompositeArrow[NodeType, ArrowType]]]:
+        """
+        Build composites at each order, until max_arrow_number is reached
+        """
+        content_before_update = set(self)
+        max_order_before_update = max(len(arr) for arr in self)
+
+        order = 1
+        while True:
+
+            added_arrows = set()
+            # loop through length n arrows and try to extend them by 1 node
+            for arr in self.arrows(
+                arrow_length_range=(order, order + 1), include_non_causal=False,
+            ):
+                for extension in self.arrows(
+                    src=arr[-1], arrow_length_range=(1, 2)):
+                    arr_candidate = arr + extension
+
+                    # verify that candidate arrow's end is causal
+                    if (arr_candidate not in self and self.get(
+                            arr_candidate[1:],
+                            default=torch.zeros(0)).sum() > 0):
+                        self.add(arr_candidate)
+                        added_arrows.add(arr_candidate)
+
+            # drop arrows above the limit
+            removed_arrows = self.prune_relations(max_arrow_number)
+
+            # if we've not looked at all arrows already in the cache
+            # continue
+            # otherwise stop as soon as nothing new comes up
+            if not (order <= max_order_before_update or any(
```

Comment:
Why stopping on this order criterion? We could envision trying to build even longer composites using the one we created? Or we really want to bound the maximum order here?

Reply:
This is what is being done. the criterion is to continue building if one of these 2 things is true:
- we didn't reach the max order of arrows already in the graph (would be kind of dumb to stop before trying to extend the longest arrow already in the graph)
- there is at least one candidate for extension at the current order (i.e with a non-zero score)

Reply:
Indeed, I misread.